### PR TITLE
Imports & WdlNamespacing - re-review

### PR DESCRIPTION
**Pull request by ruchim**
Note: There's one new test case I introduced that's broken (Namespace name collision), but I wanted to start the review, and push that fix in soon.

### CODE REVIEW

File: src/main/scala/wdl4s/Import.scala [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
+
+    Import(uri, namespaceTerminal)
   }
 }
 
 // FIXME: I dislike dragging the AST along but it's necessary for "compile" time error syntax highlighting, argh
-case class Import(uri: String, namespaceTerminal: Option[Terminal]) {
-  val namespaceName: Option[String] = namespaceTerminal.map(_.sourceString)
+case class Import(uri: String, namespaceTerminal: Terminal) {
+  val namespaceName: String = File(namespaceTerminal.sourceString).nameWithoutExtension
```
Comment (Horneth) [inline src/main/scala/wdl4s/Import.scala]: If the `namespaceTerminal` comes from the `as` clause you probably don't want/need to do `File(namespaceTerminal.sourceString).nameWithoutExtension`, although I imagine in most cases it wouldn't affect the alias string anyway.

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 2]
```diff
@@ -403,7 +414,7 @@ object WdlNamespace {
             declarationName(decl.ast), wdlType, decl.wdlType
           )))
         case Success(wdlType) =>
-          expr.evaluate(NoLookup, NoFunctions) match {
+          expr.evaluate((s: String) => throw new Throwable("not implemented"), NoFunctions) match {
```
Comment (Horneth) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Just curious, why this change ?

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 3]
```diff
@@ -124,33 +125,8 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
       """.stripMargin
   }
 
-  case object TaskAndNamespaceNameCollision extends ErrorWdl {
```
Comment (Horneth) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: So you decided to remove this one in the end ?

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 4]
```diff
@@ -610,9 +587,8 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
     CallReferencesBadInput,
     CallReferencesBadTask,
     MultipleWorkflows,
-    TaskAndNamespaceNameCollision,
     WorkflowAndNamespaceNameCollision,
-    TwoTasksHaveTheSameName,
+    //NamespaceNameCollision,
```
Comment (Horneth) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Should this be uncommented ?

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 5]
```diff
@@ -143,76 +147,82 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
...
-      scopeAst <- ast.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
-    } yield ast -> source
-
-    val topLevelScopeAsts = for {
-      namespaceAst <- nonNamespacedImports.map(_._1) :+ ast
-      scopeAst <- namespaceAst.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
-    } yield scopeAst
+    val namespaces: Seq[WdlNamespace] = for {
+      imp <- imports
+      source = tryResolve(imp.uri, importResolvers, List.empty)
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Since `tryResolve` will fail if you try to pass in 0 resolvers (because the `Nil` case requires at least one error), make the `importResolvers` parameter a `NonEmptyList`

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 6]
```diff
@@ -143,76 +147,82 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
...
-    /**
-      * For import statements with an 'as' clause, this translates to a sub-namespace
-      */
-    val namespaces: Seq[WdlNamespace] = for {
-      imp <- imports if imp.namespaceName.isDefined
-      source = importResolver(imp.uri)
-      if source.length > 0
-    } yield WdlNamespace.load(source, imp.uri, importResolver, imp.namespaceName)
+
+    def tryResolve(str: String, remainingResolvers: Seq[ImportResolver], errors: List[Throwable]): Option[WdlSource] = {
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: ToL: Add comment: `Make sure remainingResolvers has at least one entry the first time this is called`

Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: This might as well be returning `WdlSource` since we never return `None`

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 7]
```diff
@@ -143,76 +147,82 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
...
+    load(wdlSource, resource, Seq(fileResolver), None)
   }
 
   def load(wdlSource: WdlSource, resource: String, importResolver: ImportResolver): WdlNamespace = {
-    load(wdlSource, resource, importResolver, None)
+    load(wdlSource, resource, Seq(importResolver), None)
   }
 
-  private def load(wdlSource: WdlSource, resource: String, importResolver: ImportResolver, importedAs: Option[String]): WdlNamespace = {
+  private def load(wdlSource: WdlSource, resource: String, importResolver: Seq[ImportResolver], importedAs: Option[String]): WdlNamespace = {
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: `importedAs` might as well be a `String`. We should always have something to import it as now

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 8]
```diff
@@ -321,10 +331,11 @@ object WdlNamespace {
 
     def scopeNameAndTerminal(scope: Scope): (String, Terminal) = {
       scope match {
-        case ns: WdlNamespace => ("Namespace", imports.find(_.namespaceName == ns.importedAs).get.namespaceTerminal.get)
+        case ns: WdlNamespace => ("Namespace", imports.find(_.namespaceName == ns.importedAs.get).get.namespaceTerminal)
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: ToL: This'll look a lot nicer if these aren't `Option`s (as suggested above)

File: src/main/scala/wdl4s/formatter/SyntaxFormatter.scala [hunk 9]
```diff
@@ -79,8 +79,8 @@ class SyntaxFormatter(highlighter: SyntaxHighlighter = NullSyntaxHighlighter) {
   }
 
   private def formatImport(imp: Import): String = {
-    val namespace = imp.namespaceTerminal.map(ns => s" as ${ns.sourceString}").getOrElse("")
-    s"${highlighter.keyword("import")} '${imp.uri}'$namespace"
+    val namespace = imp.namespaceName
+    s"${highlighter.keyword("import")} '${imp.uri}' as $namespace"
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/formatter/SyntaxFormatter.scala]: ToL: Hmmm, this is used when highlighting source code? I think it's potentially a little confusing if we're return edited source code instead?

File: src/test/cases/three_step_import/ancestry.expectations [hunk 10]
```diff
@@ -1,10 +1,12 @@
 {
   "cgrep": [""],
-  "cgrep.count": ["cgrep", ""],
-  "cgrep.in_file": ["cgrep", ""],
-  "cgrep.pattern": ["cgrep", ""],
+  "cgrep.cgrep": ["cgrep", ""],
```
Comment (cjllanwarne) [inline src/test/cases/three_step_import/ancestry.expectations]: Scott's changes? I'll ignore all these

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 11]
```diff
@@ -173,28 +149,29 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
       """.stripMargin
   }
 
-  case object TwoTasksHaveTheSameName extends ErrorWdl {
-    val testString = "detect when two tasks have the same name"
+  //RUCHI: Needs fixing, the syntax formatter should list both import statements, not
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: TODONE?

File: src/test/scala/wdl4s/SyntaxErrorSpec.scala [hunk 12]
```diff
@@ -173,28 +149,29 @@ class SyntaxErrorSpec extends FlatSpec with Matchers {
...
-        |Task defined here (line 2, col 6):
+        |Namespace defined here (line 1, col 16):
         |
-        |task ps {
-        |     ^
+        |import "ps" as ps
+        |               ^
         |
-        |Task statement defined here (line 2, col 6):
+        |Namespace defined here (line 1, col 16):
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxErrorSpec.scala]: Presumably these two `(line 1, col 16)` references should be different?

File: src/test/scala/wdl4s/SyntaxHighlightSpec.scala [hunk 13]
```diff
@@ -163,8 +182,7 @@ class SyntaxHighlightSpec extends Matchers with WordSpecLike {
         |  call t as u {
         |    input: f="abc", p=p
         |  }
-        |}""".stripMargin,
-      importResolver = (s:String) => ""
+        |}""".stripMargin, resolver _
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxHighlightSpec.scala]: ToL: I think you can change resolver to remove this `_`
```
def resolver: String => WdlSource = (importUri: String) => {
  ...
}
```

Comment (cjllanwarne) [inline src/test/scala/wdl4s/SyntaxHighlightSpec.scala]: ToL: bring the closing `)` up?

File: src/test/scala/wdl4s/ThreeStepImportSpec.scala [hunk 14]
```diff
@@ -63,14 +63,14 @@ class ThreeStepImportSpec extends FlatSpec with Matchers {
...
+    namespace.tasks.size shouldEqual 0
   }
 
   it should "Have 0 imported WdlNamespaces" in {
-    namespace.namespaces.size shouldEqual 0
+    namespace.namespaces.size shouldEqual 3
   }
   it should "Have tasks with the names 'ps', 'cgrep' and 'wc'" in {
-    namespace.tasks map {_.name} shouldEqual Seq("ps", "cgrep", "wc")
+    namespace.namespaces flatMap {_.tasks} map {_.name} shouldEqual Seq("ps", "cgrep", "wc")
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/ThreeStepImportSpec.scala]: Misleading test name: `should "Have namespaces with the names ..."`

File: src/test/scala/wdl4s/WorkflowImportsSpec.scala [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    |
+    |}
+    |""".stripMargin
+
+  val wdlWithImports = imports + primaryWorkflow
+
+  val namespace = WdlNamespaceWithWorkflow.load(wdlWithImports, wdlDirectory)
+
+  "WDL file with imports" should "Have 0 tasks (3 tasks are in separate namespace)" in {
+    namespace.tasks.size shouldEqual 1
```
Comment (cjllanwarne) [inline src/test/scala/wdl4s/WorkflowImportsSpec.scala]: `should "Have 1 tasks...`?

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 7 above)
Comment (ruchim) [inline src/main/scala/wdl4s/WdlNamespace.scala]: @cjllanwarne Upon investigation, this optional string seem to influence how the typeEvaluator and valueEvaluator functions work -- they depend on being able to filter out WdlNamespaces with an undefined value for importedAs. So although all imports now have a namespaceName, the primary workflow still doesn't, so it truly is optional for WdlNamespace objects. This could be fixed with some restructuring...thoughts?

File: src/main/scala/wdl4s/WdlNamespace.scala (hunk 5 above)
Comment (ruchim) [inline src/main/scala/wdl4s/WdlNamespace.scala]: As mentioned above, by default we pass in atleast 1 resolver. If this method `tryResolve` is called from elsewhere, now it throws a proper message saying no import resolvers were provided.

**Comment by ruchim** [comment]
@Horneth @cjllanwarne I'm waiting for the official thumbs, please let me know if I missed any of your requests.

**Comment by Horneth** [comment]
👍

File: src/main/scala/wdl4s/WdlNamespace.scala [hunk 16]
```diff
@@ -186,17 +168,20 @@ object WdlNamespace {
       importStatement <- importAst.astListAsVector.map(Import(_))
     } yield importStatement
 
-
-    def tryResolve(str: String, remainingResolvers: Seq[ImportResolver], errors: List[Throwable]): Option[WdlSource] = {
+    //TODO: Make sure remainingResolvers has at least one entry the first time this is called
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/WdlNamespace.scala]: Since you already handle the `case Nil` below, I think this is TODONE.

**Comment by cjllanwarne** [comment]
👍
